feature(@desktop/chat) Enhance message context menu with mark as unread

kounkou opened the pull request:
[Link to status-go changes](https://github.com/status-im/status-go/pull/4382)

### Description

Feature to enhance message context menu with mark as unread.

[Figma design](https://www.figma.com/file/Mr3rqxxgKJ2zMQ06UAKiWL/%F0%9F%92%AC-Chat%E2%8E%9CDesktop?type=design&node-id=14632-460085&mode=design&t=9iVc38RXY5QVm5uz-0)

fixes #10329

### What does the PR do

The PR does the following : 

- Adds capacity to mark a message as unread
- Adds capacity to mark a message with mention as unread
- Adds persistence to the marking of the message (change can be seen at after reboot)
- Adds marking in right click contextual menu

### Affected areas

- Chat screen
- Contextual menu (Right click 

### Screenshot of functionality (including design for comparison)

[Screencast from 2023-11-30 09:08:03 AM.webm](https://github.com/status-im/status-desktop/assets/2589171/9da6b145-603d-4999-9052-228195f34cce)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 684dd9dc | [#1](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-12879/1/) | 2023-11-27 09:35:08 | ~6 min | `tests/nim` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-12879/1//consoleText) |

osmaczko reviewed (commented):
Hey, I'm just reviewing the behavior first. It seems that when you mark a message as unread, there is a no badge appearing on the chats list.

kounkou commented:
In previous commits, I was setting only 1 message as unread, I fixed the issue :

- the query in the go-status should set seen to 0 for other messages with timestamp > current_message_timestamp like @jrainville pointed out.

Now, I'll have a look again if the behavior deviates from what we want, but it should work as expected. I'll make the changes and come back with another revision.

### Edit 1 : 

I have fixed the queries.
I am seeing some inconsistencies with the app without my changes where the `@` are updated for messages without mentions (on 0.15.0-rc2)

![Screenshot from 2023-11-27 13-44-32](https://github.com/status-im/status-desktop/assets/2589171/8598eb94-d946-4d70-a22a-963279df1081)

So I am a bit confused, let's see the 3 main cases for this features, to extract what is expected :

#### case 1 : I have the latest messages in field of view and I mark a message, and the rest of messages with greater timestamp don't have mentions

- the marker is moved accordingly
- the left panel has indicators that there are some new messages. Those markers on the left panel disappear once I move the chat up and down 
- the in-chat bottom right indicators don't indicate anything because no mentions `@`

![Screenshot from 2023-11-27 13-08-19](https://github.com/status-im/status-desktop/assets/2589171/807f58f3-ce77-4cbd-9949-eb77d79c8d6c)

#### case 2 : I have the latest messages out of field of view and I mark a message, and the rest of messages with greater timestamp don't have mentions

- There's indicator that there are new messages but because there aren't mentions, there's no in chat bottom right indicator
- similar to above. We need to scroll down to reset the indicators

![Screenshot from 2023-11-27 13-15-03](https://github.com/status-im/status-desktop/assets/2589171/dc80e1c5-2512-4d4d-99dc-a5a102d586b4)

##### case 3 : I have the latest messages out of field of view and I mark a message, and the rest of the messages with greater timestamp has 1 mention

- There's indicator in left panel that there's 1 mention
- There's indicator in the in chat right bottom indicating the mention number 

![Screenshot from 2023-11-27 13-27-57](https://github.com/status-im/status-desktop/assets/2589171/9fd8631f-14ca-489b-aa50-5ae669f799f1)

osmaczko commented:
From the video, the first time you mark message as unseen, there is no badge:
![image](https://github.com/status-im/status-desktop/assets/33099791/b580ed2f-b5c2-4f02-a85b-8396ee38a193)

This is most likely because `markAllMessagesReadIfMostRecentMessageIsInViewport` is called, which effectively clears the badges. The expected behavior should be that the badge remains until you open the chat again.

osmaczko commented:
> #### case 1 : I have the latest messages in field of view and I mark a message, and the rest of messages with greater timestamp don't have mentions
> * the marker is moved accordingly
> * the left panel has indicators that there are some new messages. Those markers on the left panel disappear once I move the chat up and down
> * the in-chat bottom right indicators don't indicate anything because no mentions `@`

> Those markers on the left panel disappear once I move the chat up and down

The badge should remain until you open the chat again. To be true, I think current behavior is okeyish as well. 

> #### case 2 : I have the latest messages out of field of view and I mark a message, and the rest of messages with greater timestamp don't have mentions
> * There's indicator that there are new messages but because there aren't mentions, there's no in chat bottom right indicator
> * similar to above. We need to scroll down to reset the indicators

Same as above.
 
> ##### case 3 : I have the latest messages out of field of view and I mark a message, and the rest of the messages with greater timestamp has 1 mention
> * There's indicator in left panel that there's 1 mention
> * There's indicator in the in chat right bottom indicating the mention number

Correct.

kounkou commented:
Currently figuring out a way to produce expected behavior. The ChatMessagesView needs to be aware of the recently `mark as unread` so that the ChatMessageView doesn't clear away the badges. The badge are cleared when the chat gets `onActiveChanged` signal.

osmaczko commented on src/app/modules/main/chat_section/controller.nim:
```diff
@@ -135,6 +135,12 @@ proc init*(self: Controller) =
     self.delegate.onMarkAllMessagesRead(chat)
 
+  self.events.on(message_service.SIGNAL_MESSAGE_MARKED_AS_UNREAD) do(e:Args):
+    let args = message_service.MessageMarkMessageAsUnreadArgs(e)
+    let chat = self.chatService.getChatById(args.chatId)
+    self.chatService.updateUnreadMessage(args.chatId, args.messagesCount, args.messagesWithMentionsCount)
+    self.delegate.onMarkMessageAsUnread(chat)
```
This looks a bit like a ping-pong in between layers: `messages service` -> `controller` -> `chat service`.

Ideally, it should be: 
`messages service` -> `chat service` then
`messages service` -> `controller`

In other words, if there is no listener to `SIGNAL_MESSAGE_MARKED_AS_UNREAD` then chat service wouldn't be updated, but it should.

osmaczko replied:
In short, our architecture looks as follows:
services | controller - module - view

**services**  is the source of truth and is the place where we communicate with status-go.
**controller** is the only one who talk with services.
**module** is core logic for given module, uses controller to get/set the data, knows anything about services.
**view** is QtObject that is exposed to qml and talks only with module.

Services can talk to each other, directly. Messages service has already reference to chat service, to update the chat when it is needed, see: https://github.com/status-im/status-desktop/blob/ab7085a18424062dc0dc63c8fe55ec6d74dec7f1/src/app_service/service/message/service.nim#L293

I will try to elaborate more about architecture on Monday.

kounkou replied:
Got it, and you're right that service can talk to each others, I just thought in my mind the ask was to capture the signal  `SIGNAL_MESSAGE_MARKED_AS_UNREAD` inside the chat_service. That's the main reason why I expressed my concern.
I think I can submit another code change.
Thanks a lot @osmaczko !

osmaczko replied:
![StatusAppHLClassDiagram](https://github.com/status-im/status-desktop/assets/33099791/7ace8d64-3554-45b7-b54d-39cc7da11bef)

That explains the architecture we try to follow pretty well. The diagram itself was done by @saledjenic some time ago already. FYI: recently there was a discussion about a bit different architecture, you can follow it there https://github.com/status-im/status-desktop/pull/12489

kounkou replied:
Thanks for the share @osmaczko, this is gold 💰

osmaczko commented on ui/app/AppLayouts/Chat/views/ChatMessagesView.qml:
```diff
@@ -103,10 +108,19 @@ Item {
         }
     }
 
+    Connections {
+        target: root.messageStore
+
+        function onMessageMarkedAsUnreadChanged() {
+            d.setKeepUnread(true)
+        }
+    }
+
```
I believe this is redundant to:
```
  proc markMessageAsUnread*(self: View, messageId: string) {.slot.} =
    self.delegate.markMessageAsUnread(messageId)
    self.setKeepUnread(true)
```
and can be removed from here.

kounkou replied:
That's correct, thanks for catching that.

osmaczko commented on ui/imports/shared/views/chat/MessageContextMenuView.qml:
```diff
@@ -153,6 +154,17 @@ StatusMenu {
+        id: markMessageAsUnreadAction
+        text: qsTr("Mark as unread")
+        icon.name: "hide"
+        enabled: root.isDebugEnabled && !root.disabledForChat
```
Why only in debug mode?

kounkou replied:
Good catch, I'll update this.

osmaczko reviewed (commented):
Nice and clean, only few concerns.

osmaczko commented on test/nim/message_model_test.nim:
:muscle:

igor-sirotin commented on src/app/modules/main/chat_section/chat_content/messages/view.nim:
```diff
@@ -17,6 +17,7 @@ QtObject:
       chatIcon: string
       chatType: int
       loading: bool
+      keepUnread: bool
```
IMO `forceUnread` would be better, but it's on your choice 🙂

kounkou replied:
Thanks for the proposal, I believe `keepUnread` is more inclusive, what do you think?

igor-sirotin commented on src/app_service/service/message/service.nim:
```diff
@@ -577,6 +584,34 @@ QtObject:
     except Exception as e:
       error "error: ", procName="pinUnpinMessage", errName = e.name, errDesription = e.msg
 
+  proc markMessageAsUnread*(self: Service, chatId: string, messageId: string) = 
```
This procedure is calling a status-go endpoint that accesses the database.
In such cases we usually implement an "async task", you can find examples in neighbour file: `src/app_service/service/message/async_tasks.nim`.

Though I think we fixed lots of issues with databases, maybe it's not the case anymore?
cc @osmaczko 

_____

Also, while this procedure is syncronous, it doesn't make much sense to emit a `SIGNAL_MESSAGE_MARKED_AS_UNREAD` signal with result. We could just return it from the proc. But we would need to keep the signal for an async task ofc.

kounkou replied:
That's fair, let me update the code to have an async call.

osmaczko replied:
That's a good point, @igor-sirotin. Ideally, everything that interacts with the database should be asynchronous; nothing has changed on that topic.

igor-sirotin commented on src/app_service/service/message/service.nim:
```diff
@@ -577,6 +584,34 @@ QtObject:
+
+      if (response.result.contains("error")):
+        let errMsg = response.result["error"].getStr
+        error "error: ", procName="markMessageAsUnread", errDesription = errMsg
```
```suggestion
        error "markMessageAsUnread failed ", errDesription = errMsg
```

kounkou replied:
Updated with async code

igor-sirotin commented on src/app_service/service/message/service.nim:
```diff
@@ -734,6 +769,32 @@ QtObject:
 
     self.threadpool.start(arg)
 
+  proc onMarkMessageAsUnread*(self: Service, response: string) {.slot.} =
```
Is this function actually used now?

kounkou replied:
Thanks, removed in the latest version.

igor-sirotin reviewed (commented):
Nice! Will keep it opened until status-go PR is merged then review again.

igor-sirotin commented:
I don't see status-go update here. Please don't forget to bump status-go to include the changes from https://github.com/status-im/status-go/pull/4382.

When changes are required in both `status-desktop` and `status-go` repositories, we usually keep the branches in sync. I.e. `status-desktop/fix/terrible-issue` should point to `status-go/fix/terrible-issue`.  

1. Create a branch in status-desktop: `fix/terrible-issue`
2. Create a branch in status-go: `fix/terrible-issue`
3. When all changes are made, commit the status-go submodule change to desktop `fix/terrible-issue`.
At this point  `status-desktop/fix/terrible-issue` should point to `status-go/fix/terrible-issue`.  
4. Open 2 PRs, link them to each other.
5. Add a `waiting-for-status-go-merge` label to desktop's PR
-----
When merging, it's important for status-desktop to point to status-go `develop` branch, not your `status-go/fix/terrible-issue`. So here's the merging process:
1. Merge status-go PR
2. In desktop repo update the status-go submodule to point to the latest commit in `develop` branch (it will be your merge commit)
3. Remove the  `waiting-for-status-go-merge` label 
4. Can merge the status-desktop PR now.

igor-sirotin commented on src/app_service/service/message/async_tasks.nim:
```diff
@@ -295,3 +293,42 @@ const asyncGetMessageByMessageIdTask: Task = proc(argEncoded: string) {.gcsafe,
+
+  try:
+    let response = status_go.markMessageAsUnread(arg.chatId, arg.messageId)
+
```
I think we should also process `response.error` here. That would be the `error` returned by status-go.

kounkou replied:
Added code to process the error

igor-sirotin commented on src/app_service/service/message/service.nim:
```diff
@@ -581,6 +588,21 @@ QtObject:
+      tptr: cast[ByteAddress](asyncMarkMessageAsUnreadTask),
+      vptr: cast[ByteAddress](self.vptr),
+      slot: "onAsyncMarkMessageAsUnread",
+      requestId: $genUUID(),
```
I don't see that you actually use this `requestId` anywhere, is it needed?

kounkou replied:
Yeah, I thought we had metrics to put the request ID in, for now I'll have to remove it.

igor-sirotin reviewed (approved):
LGTM now, thanks for the async task 👍

kounkou commented:
I am currently solving merge conflicts before merging.

kounkou commented:
Reordering the commits to make commits atomic, then I'll push the change.
